### PR TITLE
CLN: assorted indexing-related cleanups

### PR DESCRIPTION
**Pull request by jbrockmendel**
- _setitem_with_indexer isnt consistent about whether or not it returns anything, make it always-None
- avoid using private loc method from DataFrame

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -862,7 +862,9 @@ def __getitem__(self, key):
 
                 return result
             except InvalidIndexError:
-                pass
+                if not isinstance(self.index, MultiIndex):
```
Comment (WillAyd) [inline pandas/core/series.py]: What is this for?

**Review by WillAyd — COMMENTED** [review]
generally looks good one question

File: pandas/core/series.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/series.py]: it wasn't obvious to me that non-MI shouldn't be allowed through here, this should clarify it for future readers (or me once I forget)

File: pandas/core/frame.py [hunk 2]
```diff
@@ -2933,12 +2933,12 @@ def __setitem__(self, key, value):
...
             self._set_item(key, value)
 
-    def _setitem_slice(self, key, value):
+    def _setitem_slice(self, key: slice, value):
         # NB: we can't just use self.loc[key] = value because that
         #  operates on labels and we need to operate positional for
         #  backwards-compat, xref GH#31469
         self._check_setitem_copy()
-        self.loc._setitem_with_indexer(key, value)
+        self.iloc[key] = value
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: `iloc` ends up calling the same method, so calling it here directly probably avoids a bunch of duplicate validation. 

So I would either leave it as is, or at least check the impact of that, or eg change to `iloc._setitem_with_indexer(..)` to make it more obvious that it is positional indexing

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -3190,17 +3190,16 @@ def is_int(v):
...
+        if is_positional and self.is_mixed():
+            try:
                 # Validate start & stop
                 if start is not None:
                     self.get_loc(start)
                 if stop is not None:
                     self.get_loc(stop)
                 is_positional = False
-        except KeyError:
-            if self.inferred_type in ["mixed-integer-float", "integer-na"]:
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/base.py]: this was not needed?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: not reachable; the check for self.is_mixed() a few lines up means these inferred_types are impossible

File: pandas/core/frame.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/frame.py]: WIll do that for now and revisit in a non-"cleanup" PR

Comment (jbrockmendel) [reply pandas/core/frame.py]: updated

**Comment by jreback** [comment]
does this change anything user facing? (no?)

**Comment by jbrockmendel** [comment]
> does this change anything user facing? (no?)

Correct, nothing user-facing.

**Review by jreback — APPROVED** [review]
lgtm. @jorisvandenbossche

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**